dashboard/app/helpers/aichat_sagemaker_helper.rb

```diff
@@ -1,62 +1,122 @@
-    all_messages.each do |msg|
-      if msg[:role] == USER
-        inputs += wrap_as_instructions(msg[:chatMessageText])
-      elsif msg[:role] == ASSISTANT
-        # Note that each assistant message in the conversation history is followed by
-        # the end-of-sentence token but a begin-of-sentence token is not required.
-        inputs += msg[:chatMessageText] + SENTENCE_END_TOKEN
+    selected_model_id = aichat_params[:selectedModelId]
+    if selected_model_id == FINE_TUNED_MODELS[:ARITHMO]
+      # Format input for Arithmo model with customized format as detailed at https://huggingface.co/akjindal53244/Arithmo-Mistral-7B.
+      inputs = "Question: " + new_message[:chatMessageText]
+      inputs = inputs + NEWLINE + "Answer:"
+    elsif selected_model_id == FINE_TUNED_MODELS[:KAREN]
+      # Format input for Karen model using ChatML. Details at https://huggingface.co/FPHam/Karen_TheEditor_V2_CREATIVE_Mistral_7B.
+      inputs = CHAT_ML_BEGIN_TOKEN + SYSTEM + NEWLINE + aichat_params[:systemPrompt] + CHAT_ML_BEGIN_TOKEN + NEWLINE
+      inputs = inputs + CHAT_ML_BEGIN_TOKEN + USER + NEWLINE + KAREN_PRETEXT + new_message[:chatMessageText] + CHAT_ML_END_TOKEN + NEWLINE
+      inputs = inputs + CHAT_ML_BEGIN_TOKEN + ASSISTANT
+    else
+      # The instruction-tuned version of Mistral accepts formatted instructions where conversation roles
+      # must start with a user prompt and alternate between user and assistant.
+      # Mistral-7B-Instruction LLM instruction format doc at https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1.
+      inputs = ""
+      if !aichat_params[:systemPrompt].empty?
+        inputs = aichat_params[:systemPrompt] + " "
+      end
+      inputs += aichat_params[:retrievalContexts].join(" ") if aichat_params[:retrievalContexts]
+      inputs = SENTENCE_BEGIN_TOKEN + wrap_as_instructions(inputs)
+      all_messages = [*stored_messages, new_message]
+      all_messages.each do |msg|
+        if msg[:role] == USER
+          inputs += wrap_as_instructions(msg[:chatMessageText])
+        elsif msg[:role] == ASSISTANT
+          # Note that each assistant message in the conversation history is followed by
+          # the end-of-sentence token but a begin-of-sentence token is not required.
+          inputs += msg[:chatMessageText] + SENTENCE_END_TOKEN
+        end
       end
     end
-
+    # Custom stopping strings used for output quality for Pirate model.
+    # Using these seem to help with latency in receiving assistant response.
+    # https://huggingface.co/phanerozoic/Mistral-Pirate-7b-v0.3
+    stopping_strings =
+      if selected_model_id == FINE_TUNED_MODELS[:PIRATE]
+        PIRATE_CUSTOM_STOPPING_STRINGS
+      else
+        []
+      end
     {
       inputs: inputs,
       parameters: {
         temperature: aichat_params[:temperature].to_f,
         max_new_tokens: MAX_NEW_TOKENS,
-        top_p: TOP_P
+        top_p: TOP_P,
+        stop: stopping_strings,
```

Comment:
Thanks @tshaffercodeorg for pointing me to this param. Seems to help with latency even if extra text still seems to be generated.